### PR TITLE
Tightening a bunch of lose ends in gateway.

### PR DESCRIPTION
**Pull request by alainjobart**
@michael-berlin I think it's all of it, let me know if I missed some fixes.

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/1910)

### CODE REVIEW

**Comment by enisoc** [comment]
Michael is back to working on this, but publishing some comments I had anyway.

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 5 of 7 files at r3, 6 of 7 files at r5, 1 of 1 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[go/vt/discovery/replicationlag.go, line 112 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHG2fu5Z_JZBzQVXQu:-KOHG2fu5Z_JZBzQVXQv:-2027768552) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/discovery/replicationlag.go#L112)):*
> ```Go
> // TrivialStatsUpdate returns true iff the old and new TabletStats
> // haven't changed enough to warrant re-calling FilterByReplicationLag.
> func TrivialStatsUpdate(old, new *TabletStats) bool {
> ```

Best not to shadow `new()` with a variable name.

---

*[go/vt/discovery/replicationlag.go, line 126 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHHE4XORbwI5RqZq6S:-KOHHE4XORbwI5RqZq6T:1821221782) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/discovery/replicationlag.go#L126)):*
> ```Go
> 	// when there is a high rep lag, it takes a long time for it to reduce,
> 	// so it is not necessary to re-calculate every time.
> 	if oldRepLag > lowRepLag && newRepLag > lowRepLag && newRepLag < oldRepLag*1.1 && newRepLag > oldRepLag*0.9 {
> ```

If `TrivialStatsUpdate()` returns true, will the `new` value be saved anyway, to become the new `old` value next time? Or will the `new` value be thrown away? Since this is key to the behavior here, it should be documented in a comment.

---

*[go/vt/vtctl/vtctl.go, line 290 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHJ9MfZV_5T0KmQ1kC:-KOHJ9MfZV_5T0KmQ1kD:-1298566718) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/vtctl/vtctl.go#L290)):*
> ```Go
> 				"Displays all of the shards in the specified keyspace."},
> 			{"WaitForDrain", commandWaitForDrain,
> 				"[-timeout <duration>] [-retry_delay <duration>] [-initial_wait <duration>] <keyspace/shard> <served tablet type>",
> ```

Don't forget to regenerate docs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1910#-:-KOHJyYuTJHzobAkHRz7:-1267413391)*

**Comment by michael-berlin** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, 1 of 1 files at r4, 7 of 7 files at r5, 1 of 1 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[go/vt/discovery/replicationlag.go, line 112 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHJO8yNUtijdlTW4LJ:-KOHJO8yNUtijdlTW4LK:484499353) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/discovery/replicationlag.go#L112)):*
> ```Go
> // TrivialStatsUpdate returns true iff the old and new TabletStats
> // haven't changed enough to warrant re-calling FilterByReplicationLag.
> func TrivialStatsUpdate(old, new *TabletStats) bool {
> ```

This looks like a good candidate for a unit test.

My other problem with this is that it's hard to tell how much it's in line with `FilterByReplicationLag` or if there are any black spots where we would classify something as trivial whereas it would not be.

---

*[go/vt/vtgate/gateway/gateway.go, line 120 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOGjuOPtn3p2Vvo3TrW:-KOGjuOPtn3p2Vvo3TrX:-2056973485) ([raw file](https://github.com/youtube/vitess/blob/5781dbad9635d46d6a2a293029dd305ee0c0e39f/go/vt/vtgate/gateway/gateway.go#L120)):*
> ```Go
> // WaitForTablets is a helper method to wait for the provided tablets,
> // up until the *initialTabletTimeout. It will log what it is doing.
> func WaitForTablets(gw Gateway, tabletTypesToWait []topodatapb.TabletType) error {
> ```

Add a comment here that it should not be confused with the method from the Gateway interface which has the same name?

Or maybe even rename it?

---

*[go/vt/vtgate/gatewaytest/grpc_discovery_test.go, line 71 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOGlJuwmZiehdZXjO31:-KOGlJuxdo0QcE88Hu5b:1356579377) ([raw file](https://github.com/youtube/vitess/blob/5781dbad9635d46d6a2a293029dd305ee0c0e39f/go/vt/vtgate/gatewaytest/grpc_discovery_test.go#L71)):*
> ```Go
> 		t.Fatalf("WaitForTablets failed: %v", err)
> 	}
> 	ctx := context.Background()
> ```

Just inline it on the next line?

---

*[go/vt/wrangler/testlib/wait_for_drain_test.go, line 56 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOGfuOOehSdUn0bfGhH:-KOGfuOOehSdUn0bfGhI:-1553794473) ([raw file](https://github.com/youtube/vitess/blob/5924212d994e8c7aa684add410e6fa047e351220/go/vt/wrangler/testlib/wait_for_drain_test.go#L56)):*
> This value needs to not be equal to -initial_wait value below.

Maybe rephrase it:

"This value must be bigger than -initial_wait below."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1910#-:-KOHKQu4TNL8UjTsA-nI:1417088721)*

**Comment by alainjobart** [comment]
Review status: 13 of 21 files reviewed at latest revision, 7 unresolved discussions.

---

*[go/vt/discovery/replicationlag.go, line 112 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHG2fu5Z_JZBzQVXQu:-KOHLBOMEwWClVttthJY:-499146009) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/discovery/replicationlag.go#L112)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Best not to shadow `new()` with a variable name.
</details>
Done.

---

*[go/vt/discovery/replicationlag.go, line 112 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHJO8yNUtijdlTW4LJ:-KOHLIFMQc9V2hGpI_L_:1391454709) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/discovery/replicationlag.go#L112)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> This looks like a good candidate for a unit test.
> 
> My other problem with this is that it's hard to tell how much it's in line with `FilterByReplicationLag` or if there are any black spots where we would classify something as trivial whereas it would not be.
</details>
Adding tests. This code was taken from our internal GW actually, so we are somewhat confident it works. 

---

*[go/vt/discovery/replicationlag.go, line 126 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHHE4XORbwI5RqZq6S:-KOHRZMxJ6U03mrdGIGY:1123946890) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/discovery/replicationlag.go#L126)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> If `TrivialStatsUpdate()` returns true, will the `new` value be saved anyway, to become the new `old` value next time? Or will the `new` value be thrown away? Since this is key to the behavior here, it should be documented in a comment.
</details>
Good point, this code has a bug, we save the new record all the time, so if the replication lag is slowing going down, we will never update.

Changed it to only save the new record if it's different.

---

*[go/vt/vtctl/vtctl.go, line 290 \[r7\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHJ9MfZV_5T0KmQ1kC:-KOHT95sO6eIylp6R2_2:-499146009) ([raw file](https://github.com/youtube/vitess/blob/3123a1bd6d1a3def5a91da735eabdcc896bc1908/go/vt/vtctl/vtctl.go#L290)):*
<details><summary><i>Previously, enisoc (Anthony Yeh) wrote…</i></summary>
> Don't forget to regenerate docs.
</details>
Done.

---

*[go/vt/vtgate/gateway/gateway.go, line 120 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOGjuOPtn3p2Vvo3TrW:-KOHRnKG7gmjRhjr66yF:-822423164) ([raw file](https://github.com/youtube/vitess/blob/5781dbad9635d46d6a2a293029dd305ee0c0e39f/go/vt/vtgate/gateway/gateway.go#L120)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Add a comment here that it should not be confused with the method from the Gateway interface which has the same name?
> 
> Or maybe even rename it?
</details>
It's the same name on purpose, it just calls the gateway method with a timeout... I'll just add a comment.

---

*[go/vt/vtgate/gatewaytest/grpc_discovery_test.go, line 71 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOGlJuwmZiehdZXjO31:-KOHSL8ZXlPYQW4dcpEv:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5781dbad9635d46d6a2a293029dd305ee0c0e39f/go/vt/vtgate/gatewaytest/grpc_discovery_test.go#L71)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> Just inline it on the next line?
</details>
Done.

---

*[go/vt/wrangler/testlib/wait_for_drain_test.go, line 56 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOGfuOOehSdUn0bfGhH:-KOHSVsUsu3_UeYoI6FU:-499146009) ([raw file](https://github.com/youtube/vitess/blob/5924212d994e8c7aa684add410e6fa047e351220/go/vt/wrangler/testlib/wait_for_drain_test.go#L56)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> > This value needs to not be equal to -initial_wait value below.
> 
> Maybe rephrase it:
> 
> "This value must be bigger than -initial_wait below."
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1910)*

**Comment by enisoc** [comment]
:lgtm:

---

Reviewed 7 of 7 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1910#-:-KOHUwSQ_BIsTeq2ENLW:1417088721)*

**Comment by michael-berlin** [comment]
:lgtm:

Two small comments.

---

Reviewed 7 of 7 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[go/vt/discovery/tablet_stats_cache_test.go, line 52 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHc3qJmMqB0GoWJyCR:-KOHc3qJmMqB0GoWJyCS:-433385931) ([raw file](https://github.com/youtube/vitess/blob/19809959670507f940cab9596fa0ad3da69ae145/go/vt/discovery/tablet_stats_cache_test.go#L52)):*
> ```Go
> 
> 	// update stats with a change that won't change health array
> 	stillHhealthyTs1 := &TabletStats{
> ```

typo: stillHealthyTs1 (extraneous `H`)

---

*[go/vt/discovery/tablet_stats_cache_test.go, line 62 \[r8\]](https://reviewable.io:443/reviews/youtube/vitess/1910#-KOHc9b1K5YIiGeybIhb:-KOHc9b1K5YIiGeybIhc:-348656224) ([raw file](https://github.com/youtube/vitess/blob/19809959670507f940cab9596fa0ad3da69ae145/go/vt/discovery/tablet_stats_cache_test.go#L62)):*
> ```Go
> 	tsc.StatsUpdate(stillHhealthyTs1)
> 
> 	// check it's there
> ```

It's worth calling out in the comment that `ts1` instead of `stillHealthyTs1` gets returned.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1910#-:-KOHcNCC4ew1MlMVNDc4:-917082244)*
